File: backend/app/rag/retriever.py
"""Retrieve relevant knowledge chunks for a user query."""

from __future__ import annotations

from dataclasses import dataclass

from app.rag.embeddings import EmbeddingProvider
from app.rag.store import SearchHit, VectorStore


@dataclass
class RetrievedContext:
    """Result of a retrieval: the relevant hits (may be empty)."""

    hits: list[SearchHit]

    @property
    def is_empty(self) -> bool:
        return not self.hits


class Retriever:
    """Embed a query, search the store, and keep hits above a score threshold."""

    def __init__(
        self,
        store: VectorStore,
        embeddings: EmbeddingProvider,
        *,
        top_k: int,
        min_score: float,
    ) -> None:
        self._store = store
        self._embeddings = embeddings
        self._top_k = top_k
        self._min_score = min_score

    @property
    def store(self) -> VectorStore:
        """Access the underlying vector store for direct operations."""
        return self._store

    async def retrieve(self, query: str) -> RetrievedContext:
        # No index yet -> behave gracefully (no embedding call, empty context).
        if self._store.size == 0 or not query.strip():
            return RetrievedContext(hits=[])
        query_vector = await self._embeddings.embed_query(query)
        hits = self._store.search(query_vector, self._top_k)
        relevant = [hit for hit in hits if hit.score >= self._min_score]
        return RetrievedContext(hits=relevant)

File: lib/rag/retriever.ts
/**
 * Retrieve relevant knowledge chunks for a user query.
 *
 * Port of `backend/app/rag/retriever.py`.
 */

import type { EmbeddingProvider } from "@/lib/rag/embeddings";
import type { SearchHit, VectorStore } from "@/lib/rag/store";

/** Result of a retrieval: the relevant hits (may be empty). */
export interface RetrievedContext {
  hits: SearchHit[];
}

export function isEmptyContext(context: RetrievedContext): boolean {
  return context.hits.length === 0;
}

/** Embed a query, search the store, and keep hits above a score threshold. */
export class Retriever {
  private readonly _store: VectorStore;
  private readonly embeddings: EmbeddingProvider;
  private readonly topK: number;
  private readonly minScore: number;

  constructor(
    store: VectorStore,
    embeddings: EmbeddingProvider,
    options: { topK: number; minScore: number },
  ) {
    this._store = store;
    this.embeddings = embeddings;
    this.topK = options.topK;
    this.minScore = options.minScore;
  }

  /** Access the underlying vector store for direct operations. */
  get store(): VectorStore {
    return this._store;
  }

  async retrieve(query: string): Promise<RetrievedContext> {
    // No index yet -> behave gracefully (no embedding call, empty context).
    if (this._store.size === 0 || !query.trim()) {
      return { hits: [] };
    }
    const queryVector = await this.embeddings.embedQuery(query);
    const hits = this._store.search(queryVector, this.topK);
    const relevant = hits.filter((hit) => hit.score >= this.minScore);
    return { hits: relevant };
  }
}
